### DISCUSSION TITLE
creating a spiral in imagemagick

### DISCUSSION THREAD

**Discussion by thisisnottaken** [ImageMagick/ImageMagick#3245] [Help]
hi,
I have seen https://legacy.imagemagick.org/Usage/transform/ where there is mention of polaroid transformation .
But , I am not able to draw a simple spiral curve .
How can I do that ? 
I can give as input the centre coordinates , the maximum radius of curve and some factor which specifies that how quickly should the curve complete a rotation and rate of increase of radius .
Is there something inbuilt for this in im or some script created by expert users ?
BTW,
I am using
[code]
magick.exe --version
Version: ImageMagick 7.0.10-50 Q16 x64 2020-12-16 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192829335
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib
[/code]
Thanks.

**Comment by snibgo** [comment]
What version of IM, on what platform? I'll assume IM v7 on Widows BAT.

Many methods can be used. For example, we regard a spiral as a polar distortion of a line. 
```
%IMG7%magick ^
  -size 500x100 xc:White ^
  -stroke Black -draw "line 0,99,499,0" ^
  +write spiral_base.png ^
  +distort polar -1 ^
  spiral_out.png
```
We don't need to write spiral_base.png, but it helps us to understand:
![spiral_base](https://user-images.githubusercontent.com/33812368/107946693-4abc7700-6f89-11eb-98ce-d7024d7f2aa6.png)

spiral_out.png is:
![spiral_out](https://user-images.githubusercontent.com/33812368/107946719-51e38500-6f89-11eb-838b-617acbfacb3a.png)

**Reply by thisisnottaken** [reply]
I want more circles in it . 
How to get more rotations in the above curve ?

Thanks.

**Comment by GeeMack** [comment]
As **snibgo** mentioned, there are several ways to create a spiral using ImageMagick. This example lets you easily choose how many "rings" you want and the spacing between them.

    magick -size 640x4 xc:black ^
       -background none -splice 0x1 -extent 0x36 ^
       -duplicate 8 +delete -append +repage ^
       -distort affine "0,0 0,0 0,1 0,1 %[w],0 %[w],36" ^
       -distort arc "360 0 %[h] 0" -shave 1 +repage ^
       -background white -flatten spiral.png

That creates this image...

![spiral](https://user-images.githubusercontent.com/65475279/108022872-08c21e00-6fe7-11eb-8a6b-288bc817c4c0.png)

The thickness of the line is set by the height specified in `-size 640x4`, so here it's 4 pixels. The distance from the top of one "ring" to the top of the next is set by the height specified in `-extent 0x36`, making it 36 pixels here. The number of "rings" is determined by the argument in `-duplicate 8`, so there will be 8 "rings" going from the bottom edge to the center.

The command starts by creating a black rectangle 640 by 4 pixels, adds a single transparent row to the top, and extends it with a transparent background to 36 pixels high. That stripe is duplicated to make the desired number of "rings", and all of those are appended vertically to make a block of stripes.

Next the `-distort affine` operation more or less slides the far right edge of the image downward by the amount you specified for the spacing from "ring" to "ring". Make sure the very last argument in the `-distort affine` operation matches the amount you specified in the preceding `-extent` operation.

Then the `-distort arc` operation wraps that image around 360 degrees to complete the spiral. That operation adds a single row of pixels all the way around, so we use `-shave 1` to remove it. The inner end of the spiral line should be almost exactly in the center of the image.

The output here will be on a white background. If you want a transparent background just omit the `-background white -flatten` at the end. 

This should work in almost any version of IM from 6.8 to current by using `convert` for 6x versions and `magick` for 7x versions. Change the continued-line carets "^" to backslashes "\\" to make it work in a *nix shell.

**Reply by thisisnottaken** [reply]
Thanks .
That is a very nice explanation.

One more thing - if you compare the eight turn spiral to single turn spiral by snibgo then it seems that in your case the radius of spiral is not changing continuously . I think you are combining semi circles whose centre have been shifted.

I am looking for eight turn spiral version of snibgo. Is it possible to get that ?

**Comment by thisisnottaken** [comment]
Geemack - in your spiral the distance between turns is constant .
I am looking for logarithmic spiral in which distance between consecutive turns increases.
See https://www.britannica.com/science/spiral-mathematics .

Can you suggest for this specific requirement ?

Thanks.

**Comment by snibgo** [comment]
The commands by GeeMack and me create Archimedean spirals. They both work by applying a polar transformation to a straight line. We can make a logarithmic spiral by applying a polar transformation to an exponential curve.

IM does not contain simple facilities for creating curves from mathematical functions. gnuplot is a better program for that. But it can be done in IM. For example:
```
%IMG7%magick ^
  -size 1x1000 gradient: -rotate 90 ^
  -evaluate exp 1 ^
  -auto-level ^
  x.png

call %PICTBAT%graphLineCol x.png . . 0 x_glc.png

%IMG7%magick ^
  x_glc.png -negate ^
  -crop 8x1@+0+0 +repage ^
  -background None ^
  -compose Darken -layers Flatten ^
  +write e.png ^
  +write spiral_basee.png ^
  +distort polar -1 ^
  spiral_oute.png
```
![spiral_oute](https://user-images.githubusercontent.com/33812368/108062865-4b6d1000-7052-11eb-9a77-400739505fab.png)

The script graphLineCol.bat is shown at [Clut cookbook](http://im.snibgo.com/ckbkClut.htm#graphLineCol.bat).

EDIT: For an expanding interval, use eg "-evaluate Pow 0.3".

EDIT2: To get rid of the dark smudge, insert "-virtual-pixel White" before "+distort polar -1".

**Reply by thisisnottaken** [reply]
yes I am looking for spiral in which lines go father and farther away from each other.

what is  %PICTBAT% here ?
BTW , I am using windows OS.

**Comment by snibgo** [comment]
%PICTBAT% is the path to the directory where I keep image processing BAT scripts.

Similarly, %IMG7% is the path to the directory of IM executables.

**Reply by thisisnottaken** [reply]
Thank you.
That is a pretty complex script.
You are an expert in this field of image processing.
I saved graphLineCol.bat in current folder.

Though there was an error after running
call graphLineCol x_03.png . . 0 x_glc_03.png

`'echoRestore' is not recognized as an internal or external command,`
Maybe that function is in another batch file.
If possible, please point out the missing part for future readers of this thread.

But I got the output as you have shown.

For expanding rings I got
![spiral_oute_03](https://user-images.githubusercontent.com/19471922/108454926-55318600-7293-11eb-9164-ea14c74a8fcf.png)

Thanks a lot.

**Comment by snibgo** [comment]
All my published BAT scripts are available in a single ZIP file: http://im.snibgo.com/imsnibgoBats.zip . (This also includes C code for my published filters.)

The scripts are described at [Zipped BAT files](http://im.snibgo.com/zipbats.htm).

A index to my published pages about image processing with ImageMagick is at [Index](http://im.snibgo.com/index.htm).
